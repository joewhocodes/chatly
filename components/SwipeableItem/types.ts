export type SwipeableItemProps = {
    item: {
      chatName: string;
      chatId: string;
      messages: any[];
    };
    onDelete: (id: string) => void;
    onPress: () => void;
};