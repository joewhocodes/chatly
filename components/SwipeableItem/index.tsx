// SwipeableItem.tsx
import React, { useRef, useState } from 'react';
import { TouchableOpacity, Animated } from 'react-native';
import { Text } from 'native-base';
import { Swipeable } from 'react-native-gesture-handler';

import { SwipeableItemProps } from './types';

const SwipeableItem = ({ item, onDelete, onPress }: SwipeableItemProps) => {
	const swipeableRef = useRef<Swipeable>(null);
	const [isSwiping, setIsSwiping] = useState(false);

	const handleSwipeStart = () => {
		setIsSwiping(true);
	};

	const handleSwipeRelease = () => {
		setIsSwiping(false);
	};

	const handlePress = () => {
		if (!isSwiping) {
			onPress();
		}
	};

	const renderLeftActions = (
		progress: Animated.AnimatedInterpolation<number>,
		dragX: Animated.AnimatedInterpolation<number>
	) => {
		const onDeletePress = () => {
			onDelete(item.chatId);
			setIsSwiping(false);
		};

		const trans = dragX.interpolate({
			inputRange: [0, 100],
			outputRange: [0, 1],
			extrapolate: 'clamp',
		});

		return (
			<TouchableOpacity onPress={onDeletePress}>
				<Animated.View
					style={{
						backgroundColor: 'red',
						justifyContent: 'center',
						alignItems: 'flex-end',
						padding: 20,
						opacity: trans,
					}}>
					<Text style={{ color: 'white' }}>Delete</Text>
				</Animated.View>
			</TouchableOpacity>
		);
	};

	return (
		<Swipeable
			ref={swipeableRef}
            
			renderLeftActions={renderLeftActions}
			onSwipeableWillOpen={handleSwipeStart}
			onSwipeableWillClose={handleSwipeRelease}>
			<TouchableOpacity
				activeOpacity={1}
				onPress={handlePress}
				style={{ backgroundColor: 'white', padding: 20 }}>
				<Text fontFamily={'Jua-Regular'} textAlign={'center'}>
					{item.chatName}
				</Text>
			</TouchableOpacity>
		</Swipeable>
	);
};

export default SwipeableItem;
