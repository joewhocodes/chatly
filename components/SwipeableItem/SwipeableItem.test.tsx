import React from 'react';
import renderer from 'react-test-renderer';
import { NativeBaseProvider } from 'native-base';
import { render } from '@testing-library/react-native';

import SwipeableItem from './index';

jest.mock('react-native-gesture-handler', () => {
    return {
      Swipeable: jest.fn(),
    };
  });
  
describe('SwipeableItem', () => {
	it('renders SwipeableItem correctly', () => {
		const item = {
            chatName: 'Test Name',
            chatId: 'Test Id',
            messages: [],
        }
		const onDelete = jest.fn();
		const onPress = jest.fn();

		const tree = renderer
			.create(
				<NativeBaseProvider>
					<SwipeableItem
                        item={item}
                        onDelete={onDelete}
                        onPress={onPress}
					/>
				</NativeBaseProvider>
			)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
