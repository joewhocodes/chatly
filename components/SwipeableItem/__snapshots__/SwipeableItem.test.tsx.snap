// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeableItem renders SwipeableItem correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;
